all: switch from merkle tree to serial hash

tuxcanfly opened the pull request:
This is a protocol level change which switches away from the random-write and merkle tree based approach to a append-only serial hash method.

The main benefit of this switch is the simplified sync protocol and less code complexity. As an example, the sync can be done in fewer messages and larger chunks of sectors now that the blob is guaranteed to retain data written earlier as it is now append-only.

N̶o̶t̶e̶ ̶t̶o̶ ̶r̶e̶v̶i̶e̶w̶e̶r̶s̶:̶ ̶I̶ ̶h̶a̶v̶e̶ ̶t̶r̶i̶e̶d̶ ̶t̶o̶ ̶k̶e̶e̶p̶ ̶t̶h̶i̶s̶ ̶c̶h̶a̶n̶g̶e̶s̶e̶t̶ ̶a̶s̶ ̶c̶o̶m̶p̶a̶c̶t̶ ̶a̶s̶ ̶p̶o̶s̶s̶i̶b̶l̶e̶ ̶h̶o̶w̶e̶v̶e̶r̶,̶ ̶s̶o̶m̶e̶ ̶u̶n̶r̶e̶l̶a̶t̶e̶d̶ ̶c̶h̶a̶n̶g̶e̶s̶ ̶m̶i̶g̶h̶t̶ ̶h̶a̶v̶e̶ ̶c̶r̶e̶p̶t̶ ̶i̶n̶.̶ ̶P̶l̶e̶a̶s̶e̶ ̶i̶g̶n̶o̶r̶e̶ ̶a̶n̶y̶t̶h̶i̶n̶g̶ ̶t̶h̶a̶t̶ ̶d̶o̶e̶s̶ ̶n̶o̶t̶ ̶l̶o̶o̶k̶ ̶r̶e̶l̶e̶v̶a̶n̶t̶ ̶t̶o̶ ̶t̶h̶e̶ ̶t̶o̶p̶i̶c̶.̶ ̶A̶l̶s̶o̶,̶ ̶I̶ ̶h̶a̶v̶e̶ ̶n̶o̶t̶ ̶m̶o̶d̶i̶f̶i̶e̶d̶ ̶t̶h̶e̶ ̶e̶x̶e̶c̶u̶t̶a̶b̶l̶e̶s̶ ̶a̶n̶d̶ ̶i̶n̶s̶t̶e̶a̶d̶ ̶a̶d̶d̶e̶d̶ ̶a̶ ̶n̶e̶w̶ ̶p̶a̶i̶r̶ ̶(̶`̶d̶d̶r̶p̶d̶`̶,̶ ̶`̶d̶d̶r̶p̶c̶l̶i̶`̶)̶ ̶i̶n̶s̶t̶e̶a̶d̶.̶ ̶S̶o̶m̶e̶ ̶o̶f̶ ̶t̶h̶e̶ ̶t̶e̶s̶t̶s̶ ̶a̶r̶e̶ ̶f̶a̶i̶l̶i̶n̶g̶ ̶a̶n̶d̶ ̶w̶i̶l̶l̶ ̶b̶e̶ ̶f̶i̶x̶e̶d̶ ̶b̶e̶f̶o̶r̶e̶ ̶f̶i̶n̶a̶l̶i̶z̶i̶n̶g̶ ̶t̶h̶i̶s̶ ̶P̶R̶.̶

Converted to Draft and added a pending list of TODO items before this is ready.

tuxcanfly committed:
all: switch from merkle tree to serial hash

mslipper commented:
Please change the package name back to `fnd`, and remove references to `fnd.localhost`.

mslipper commented:
We also don't need to vendor dwire anymore.

tuxcanfly commented:
Ok I will rename it back to `fnd`.  I would need to vendor dwire though, as I need to bump up the `DefaultMaxByteFieldLen`, which is required for `BlobRes.Payload` in the worst case.

tuxcanfly committed:
replace github.com/ddrp-org/ddrp -> fnd

tuxcanfly committed:
rename ddrpd -> fnd

tuxcanfly commented:
TODO:

- [x] Fix tests around blob io
- [ ] B̶u̶m̶p̶ ̶u̶p̶ ̶M̶a̶x̶ ̶B̶l̶o̶b̶ ̶S̶i̶z̶e̶ ̶(̶1̶M̶B̶ ̶-̶>̶ ̶8̶M̶B̶?̶)̶
- [x] Clean up stray references to ddrp/fnd and vice-versa
- [x] Tidy up diff

Not touching blob size at this time.

mslipper commented on go.mod:
```diff
@@ -2,23 +2,17 @@ module fnd
-	github.com/golang/protobuf v1.4.2
+	github.com/golang/protobuf v1.4.0-rc.3
 	github.com/golang/snappy v0.0.1 // indirect
+	github.com/google/go-cmp v0.5.0 // indirect
```
Please revert changes to go.mod that downgrade software versions.

mslipper commented on protocol/ban_list.go:
```diff
@@ -22,8 +22,8 @@ func ParseBanListVersion(line string) (int, error) {
 	}
-	if splits[0] != "FNBAN" {
-		return 0, errors.New("ban list version must start with FNBAN")
+	if splits[0] != "DDRPBAN" {
```
Revert

mslipper commented on protocol/ban_list_test.go:
```diff
@@ -13,11 +13,11 @@ func TestParseBanListVersion(t *testing.T) {
-		{"FNBAN:beep", "end with v followed by a digit"},
-		{"FNBAN:1", "end with v followed by a digit"},
+		{":", "start with DDRPBAN"},
+		{"DDRPBAN", "colon-separated components"},
```
Revert whole file

mslipper commented on protocol/epoch_test.go:
```diff
@@ -0,0 +1,280 @@
+	}{
+		{
+			"syncs sectors when the local node has never seen the name before",
+			func(t *testing.T, setup *epochTestSetup) {
```
These tests belong somewhere else, since they're not testing the very simple epoch method defined in `epoch.go`.

tuxcanfly replied:
Moved to `updater_test.go`

mslipper commented on protocol/epoch_test.go:
```diff
@@ -0,0 +1,280 @@
+			},
+		},
+	}
+	for _, tt := range tests {
```
Need simple tests for the `epoch.go` file above.

tuxcanfly replied:
Added simple test case in `epoch_test.go`

mslipper commented on protocol/name_importer.go:
```diff
@@ -213,15 +214,15 @@ func ExtractTXTRecordsBlock(block *primitives.Block) []*FNRecord {
-func ExtractTXTRecordsTx(tx *primitives.Transaction) []*FNRecord {
-	var out []*FNRecord
+func ExtractTXTRecordsTx(tx *primitives.Transaction) []*DDRPKeyRecord {
+	var out []*DDRPKeyRecord
```
Revert whole file

mslipper commented on protocol/name_importer_test.go:
```diff
@@ -3,29 +3,30 @@ package protocol
-func TestParseFNRecord(t *testing.T) {
+func TestParseDDRPKeyRecord(t *testing.T) {
 	invalid := []string{
-		"f000000000000000000000000000000000000000000000000000000000000000000",
```
Revert whole file

mslipper commented on scripts/release-public.sh:
```diff
@@ -8,16 +8,16 @@ tag="$(git describe --tags --abbrev=0)"
   mv -f "$build_dir/fnd-cli-$1-amd64" "$build_dir/fnd-cli"
-  tar -czvf "$build_dir/fnd-$version-$1-amd64.tgz" -C "$build_dir" "fnd" "fnd-cli"
-  gpg2 --detach-sig --default-key D4B604F1 --output "$build_dir/fnd-$version-$1-amd64.tgz.sig" "$build_dir/fnd-$version-$1-amd64.tgz"
+  tar -czvf "$build_dir/ddrp-$version-$1-amd64.tgz" -C "$build_dir" "fnd" "fnd-cli"
```
Revert whole file

mslipper commented on testutil/testflags/flags.go:
```diff
@@ -6,15 +6,15 @@ import (
 
 func IntegrationTest(t *testing.T) {
-	_, ok := os.LookupEnv("Footnote_ENABLE_INTEGRATION_TESTS")
+	_, ok := os.LookupEnv("DDRP_ENABLE_INTEGRATION_TESTS")
```
FND_ENABLE_*

mslipper commented on testutil/testflags/flags.go:
```diff
@@ -6,15 +6,15 @@ import (
 
 func HandshakeTest(t *testing.T) {
-	_, ok := os.LookupEnv("Footnote_ENABLE_HANDSHAKE_TESTS")
+	_, ok := os.LookupEnv("DDRP_ENABLE_HANDSHAKE_TESTS")
```
FND_ENABLE_*

mslipper commented on testutil/testfs/fs.go:
```diff
@@ -8,15 +8,15 @@ import (
 	}
 }
 
 func NewTempFile(t *testing.T) (*os.File, func()) {
```
revert whole file

mslipper commented on blob/blob.go:
```diff
@@ -41,6 +45,14 @@ func (b *blobImpl) Name() string {
 	return b.name
 }
 
+func (b *blobImpl) Seek(sectorSize uint16) {
```
Remove `at` and make `Seek` conform to the `io.Seeker` interface.

tuxcanfly replied:
Done.

mslipper commented on blob/seal.go:
```diff
@@ -2,23 +2,26 @@ package blob
+func SealHash(name string, epochHeight, sectorSize uint16, sectorTipHash crypto.Hash, reservedRoot crypto.Hash) crypto.Hash {
 	h, _ := blake2b.New256(nil)
-	if _, err := h.Write([]byte("FNBLOB")); err != nil {
+	if _, err := h.Write([]byte("DDRPBLOB")); err != nil {
```
FNBLOB

mslipper commented on blob/store.go:
```diff
@@ -91,6 +92,14 @@ func (w *wrappedBlob) Name() string {
 	return w.blob.Name()
 }
 
+func (w *wrappedBlob) Seek(sectorSize uint16) {
```
Conform to `io.Seeker`

tuxcanfly replied:
Will do that.

tuxcanfly replied:
Done.

mslipper commented on cmd/fnd/cmd/start.go:
```diff
@@ -45,7 +47,7 @@ var startCmd = &cobra.Command{
 		lgr := log.WithModule("main")
 
-		lgr.Info("starting fnd", "git_commit", version.GitCommit, "git_tag", version.GitTag)
+		lgr.Info("starting ddrp", "git_commit", version.GitCommit, "git_tag", version.GitTag)
```
fnd

mslipper commented on protocol/syncer.go:
```diff
@@ -1,267 +1,146 @@
-					lgr.Trace("received unsolicited sector", "sector_id", msg.SectorID, "peer_id", peerID)
-					continue
+				var sectorTipHash crypto.Hash = opts.PrevHash
+				for i := 0; int(i) < len(msg.Payload); i++ {
```
Do you need to cast `int` here?

mslipper commented on protocol/syncer.go:
```diff
@@ -1,267 +1,146 @@
-				if err := opts.Tx.WriteSector(msg.SectorID, msg.Sector); err != nil {
-					lgr.Error("failed to write sector", "sector_id", msg.SectorID, "err", err)
-					continue
+				for i := 0; int(i) < len(msg.Payload); i++ {
```
Ditto

mslipper commented on protocol/syncer.go:
```diff
@@ -1,267 +1,146 @@
+		case <-payloadProcessedCh:
+			lgr.Debug("payload processed")
+			break payloadLoop
+		case <-time.NewTimer(opts.Timeout).C:
```
Timeout is being redefined with every iteration of the for loop - is this expected?

tuxcanfly replied:
Moved it out of the loop.

mslipper commented on protocol/update_queue.go:
```diff
@@ -102,65 +102,58 @@ func (u *UpdateQueue) Enqueue(peerID crypto.Hash, update *wire.Update) error {
 	}
 
-	if entry.Timestamp.After(update.Timestamp) {
+	if entry.SectorSize > update.SectorSize {
```
Should probably rename the error if it's a misnomer now.

mslipper commented on protocol/update_queue.go:
```diff
@@ -102,65 +102,58 @@ func (u *UpdateQueue) Enqueue(peerID crypto.Hash, update *wire.Update) error {
+	if entry.SectorSize > update.SectorSize {
 		return ErrUpdateQueueStaleTimestamp
 	}
 	if entry.Signature != update.Signature {
```
Is this still possible?

mslipper commented:
Given the number of "cruft" changes (like reordering imports and renaming things) it's pretty hard for me to review the parts of this PR that are material to the protocol. Let's walk through it on Monday/Tuesday when we meet.

tuxcanfly commented:
Sorry about that I didn't have much time to clean up when copying over the changes from my non-git codebase, unfortunately. I will clean up and tidy things so we can review the changes that are actually important.

tuxcanfly committed:
all: cleanup

tuxcanfly committed:
blob: fix seal prefix

tuxcanfly committed:
testutil: rename test prefix

tuxcanfly committed:
testutil: revert flags

tuxcanfly committed:
dwire: bump MaxByteFieldLen

tuxcanfly committed:
protocol: update test urls; fix parse name record

tuxcanfly committed:
blob, p2p: update and fix tests

tuxcanfly committed:
p2p, wire: update test fixtures; fix handshake test

tuxcanfly committed:
syncer: reuse timeout

tuxcanfly committed:
protocol: refactor and separate epoch/updater tests

tuxcanfly commented:
Cleaned up the code and fixed some failing tests. Ideally would have preferred the fixes to go in another branch. Working on addressing the review items next.

tuxcanfly committed:
blob: switch sectorlen, sectorcount consts

tuxcanfly committed:
blob: use io.Seeker interface in Transaction

tuxcanfly committed:
blob: seeker - additional checks

tuxcanfly committed:
blob: cleanup

tuxcanfly committed:
all: rename merkle root -> sector tip hash

tuxcanfly committed:
epoch: fix test

tuxcanfly committed:
blob: fix uint16 overflow in tx.Seek

tuxcanfly committed:
protocol: fix issue with dequeue

tuxcanfly committed:
epoch: update test

tuxcanfly committed:
vendor: use float64 for ChainInfo.progress instead of int

tuxcanfly committed:
epoch: update test to use mock time

tuxcanfly committed:
updater: update epoch checks

tuxcanfly committed:
deps: rm mstream

tuxcanfly committed:
dwire: bump max var array len

tuxcanfly committed:
protocol: rename errors to match sector instead of timestamp

tuxcanfly committed:
syncer: write equivocation proof to level store

tuxcanfly committed:
cli: blob write - add flag to reset epoch

tuxcanfly committed:
rpc: add reset epoch method placeholder

tuxcanfly committed:
rpc: implement reset epoch; reset blob

tuxcanfly committed:
rpc: blob writer - use reset epoch

tuxcanfly committed:
WIP: move signature from update to blob_res

tuxcanfly committed:
WIP: validate blob res signature - TODO fix tests

signature validation was moved from update to blob res

signature covers all the update item like sector tip hash, sector
size, epoch height, reserved root.

With the new approach, the sector tip hash is recomputed, then
signature is validated. Same with sector size, it's recomputed as
payload position + len(payload). Ideally, all information in the blob
res is enough to validate itself. i.e. prev hash, payload pos, payload.

tests are failing because updater does not expect to sign over all the
data in blob res. sector server needs to be updated to serve a valid
signature.

tuxcanfly committed:
protocol: update tests to validate blob res

since signature is now verified on blob res rather than update,
the tests needed to be updated to handle this. Mostly this is just
a matter of setting up the header for the name, so that the syncer
can use it to fetch the pubkey and match the signature for the payload.

If anything breaks in sync, the most likely culprit is the last couple
of commits that rework the syncing mechanism to switch validating
updates to validating blob res.

tuxcanfly committed:
syncer: verify update prev hash

tuxcanfly committed:
protocol: handle prev hash mismatch

tuxcanfly committed:
test: restore invalid payload signature test

tuxcanfly committed:
SectorLen->SectorBytes; SectorCount->MaxSectors

tuxcanfly committed:
test: add docs

tuxcanfly committed:
syncer: verify epoch height, prev hash after sig

tuxcanfly committed:
syncer: verify prev hash and check for equivocation

tuxcanfly committed:
syncer: minor - break after err equivocation

tuxcanfly committed:
all: generate equivocation proof

tuxcanfly committed:
protocol: handle equivocation proof request

tuxcanfly committed:
protocol: handle equivocation update

tuxcanfly committed:
protocol: handle equivocation response

tuxcanfly committed:
protocol: validateBlobRes - pass db instead of opts

tuxcanfly committed:
protocol: fix update queue tests

tuxcanfly committed:
protocol: simplify validation

tuxcanfly committed:
protocol: add TODOs

tuxcanfly committed:
wire: update equivocation proof message

tuxcanfly committed:
protocol: update equivocation proof construction

tuxcanfly committed:
protocol: update equivocation proof test

tuxcanfly committed:
protocol: handle equivocation request

tuxcanfly committed:
protocol: rename equivocation proof fields

tuxcanfly committed:
protocol: update TODO

tuxcanfly committed:
protocol: validate eq proof

tuxcanfly committed:
protocol: handle valid equivocation proof

tuxcanfly committed:
protocol: epoch height and equivocation status check

tuxcanfly committed:
protocol: set header banned time and skip if so

tuxcanfly committed:
updater: set sector tip hash on header

tuxcanfly committed:
write: fix reset epoch flag handler

tuxcanfly committed:
protocol: refactor syncsector to return signature

tuxcanfly committed:
protocol: SyncSectors return validated metadata

tuxcanfly committed:
syncer: err if already equivocated

tuxcanfly committed:
protocol: skip updates if name is equivocated

tuxcanfly committed:
rpc: return header ban time in blob list cli

tuxcanfly committed:
all: rename vars, add docs

tuxcanfly committed:
cli: update blob info
